// @vuepress/cli/dist/index.js
import { createRequire } from "node:module";
import process from "node:process";
import { DEFAULT_IGNORE_PATTERNS, createBuildApp, createDevApp, createPage, preparePageChunk, prepareRoutes } from "@vuepress/core";
import { colors, debug, formatMs, fs, hash, importFileDefault, isChildPath, logger, ora, path, picomatch, withSpinner } from "@vuepress/utils";
import { pathToFileURL } from "node:url";
import { build as build$1 } from "esbuild";
import { ensureEndingSlash, ensureLeadingSlash } from "@vuepress/shared";
import chokidar from "chokidar";
import envinfo from "envinfo";
import { cac } from "cac";
//#region src/config/loadUserConfig.ts
/**
* Load user config file
*/
const loadUserConfig = async (userConfigPath) => {
	if (!userConfigPath) return {
		userConfig: {},
		userConfigDependencies: []
	};
	const dirnameVarName = "__vite_injected_original_dirname";
	const filenameVarName = "__vite_injected_original_filename";
	const importMetaUrlVarName = "__vite_injected_original_import_meta_url";
	const result = await build$1({
		absWorkingDir: process.cwd(),
		entryPoints: [userConfigPath],
		write: false,
		target: [`node${process.versions.node}`],
		platform: "node",
		bundle: true,
		format: "esm",
		mainFields: ["main"],
		sourcemap: "inline",
		metafile: true,
		define: {
			"__dirname": dirnameVarName,
			"__filename": filenameVarName,
			"import.meta.url": importMetaUrlVarName,
			"import.meta.dirname": dirnameVarName,
			"import.meta.filename": filenameVarName
		},
		plugins: [{
			name: "externalize-deps",
			setup(pluginBuild) {
				pluginBuild.onResolve({ filter: /.*/ }, ({ path: id }) => {
					if (!id.startsWith(".") && !path.isAbsolute(id)) return { external: true };
					return null;
				});
			}
		}, {
			name: "inject-file-scope-variables",
			setup(pluginBuild) {
				pluginBuild.onLoad({ filter: /\.[cm]?[jt]s$/ }, async (args) => {
					const contents = await fs.readFile(args.path, "utf-8");
					const injectValues = `const ${dirnameVarName} = ${JSON.stringify(path.dirname(args.path))};const ${filenameVarName} = ${JSON.stringify(args.path)};const ${importMetaUrlVarName} = ${JSON.stringify(pathToFileURL(args.path).href)};`;
					return {
						loader: args.path.endsWith("ts") ? "ts" : "js",
						contents: injectValues + contents
					};
				});
			}
		}]
	});
	const { text } = result.outputFiles[0];
	const tempFilePath = `${userConfigPath}.${hash(text)}.mjs`;
	let userConfig;
	try {
		await fs.writeFile(tempFilePath, text);
		userConfig = await importFileDefault(tempFilePath);
	} finally {
		await fs.rm(tempFilePath);
	}
	return {
		userConfig,
		userConfigDependencies: Object.keys(result.metafile.inputs)
	};
};
//#endregion
//#region src/config/resolveAppConfig.ts
/**
* Resolve app config according to:
*
* - default options
* - user config file
* - cli options
*/
const resolveAppConfig = ({ defaultAppConfig, cliAppConfig, userConfig }) => {
	const appConfig = {
		...defaultAppConfig,
		...userConfig,
		...cliAppConfig
	};
	if (appConfig.source === void 0) {
		logger.error(`The ${colors.magenta("source")} option is missing.`);
		return null;
	}
	if (appConfig.bundler === void 0 || appConfig.theme === void 0) {
		logger.error(`The ${colors.magenta("bundler")} or ${colors.magenta("theme")} option is missing. For more details: ${colors.green("https://vuepress.vuejs.org/guide/troubleshooting.html#the-bundler-theme-option-is-missing")}`);
		return null;
	}
	if (appConfig.base && !(appConfig.base.startsWith("/") && appConfig.base.endsWith("/"))) {
		const rawBase = appConfig.base;
		appConfig.base = ensureLeadingSlash(ensureEndingSlash(rawBase));
		logger.warn(`${colors.magenta("base")} should start and end with a slash (/), so it has been normalized from ${colors.magenta(rawBase)} to ${colors.magenta(appConfig.base)}`);
	}
	if (appConfig.dest && isChildPath(appConfig.source, appConfig.dest)) {
		logger.warn(`${colors.magenta("dest")} directory would be emptied during build, so we fallback it to the default directory for the safety of your source files`);
		delete appConfig.dest;
	}
	return appConfig;
};
//#endregion
//#region src/config/resolveCliAppConfig.ts
const OPTIONS_COMMON = [
	"debug",
	"open",
	"port",
	"host"
];
const OPTIONS_DIRECTORY = [
	"cache",
	"dest",
	"temp"
];
/**
* Resolve app config according to command options of cli
*/
const resolveCliAppConfig = (sourceDir, cliOptions, cwd = process.cwd()) => {
	const appConfig = { source: path.resolve(cwd, sourceDir) };
	OPTIONS_COMMON.forEach((name) => {
		if (cliOptions[name] !== void 0) appConfig[name] = cliOptions[name];
	});
	OPTIONS_DIRECTORY.forEach((name) => {
		if (cliOptions[name] !== void 0) appConfig[name] = path.resolve(cwd, cliOptions[name]);
	});
	return appConfig;
};
//#endregion
//#region src/config/resolveUserConfigConventionalPath.ts
/**
* Resolve conventional user config file path
*/
const resolveUserConfigConventionalPath = (source, cwd = process.cwd()) => [
	path.resolve(cwd, "vuepress.config.ts"),
	path.resolve(cwd, "vuepress.config.js"),
	path.resolve(cwd, "vuepress.config.mjs"),
	path.resolve(source, ".vuepress/config.ts"),
	path.resolve(source, ".vuepress/config.js"),
	path.resolve(source, ".vuepress/config.mjs")
].find((item) => fs.pathExistsSync(item));
//#endregion
//#region src/config/resolveUserConfigPath.ts
/**
* Resolve file path of user config
*/
const resolveUserConfigPath = (config, cwd = process.cwd()) => {
	const configPath = path.resolve(cwd, config);
	if (!fs.pathExistsSync(configPath)) throw logger.createError(`config file does not exist: ${colors.magenta(config)}`);
	return configPath;
};
//#endregion
//#region src/config/resolveUserConfig.ts
const log$2 = debug("vuepress:cli/resolveUserConfig");
const resolveUserConfig = async ({ cliOptionsConfig, cliAppConfigSource }) => {
	const userConfigPath = cliOptionsConfig ? resolveUserConfigPath(cliOptionsConfig) : resolveUserConfigConventionalPath(cliAppConfigSource);
	log$2(`userConfigPath:`, userConfigPath);
	const { userConfig, userConfigDependencies } = await loadUserConfig(userConfigPath);
	return {
		userConfig,
		userConfigDependencies,
		userConfigPath
	};
};
//#endregion
//#region src/config/transformUserConfigToPlugin.ts
/**
* Transform user config to a vuepress plugin
*/
const transformUserConfigToPlugin = (userConfig, source, cwd = process.cwd()) => {
	const userConfigPlugin = {
		name: "user-config",
		...userConfig
	};
	userConfigPlugin.clientConfigFile ??= [
		path.resolve(cwd, "vuepress.client.ts"),
		path.resolve(cwd, "vuepress.client.js"),
		path.resolve(cwd, "vuepress.client.mjs"),
		path.resolve(source, ".vuepress/client.ts"),
		path.resolve(source, ".vuepress/client.js"),
		path.resolve(source, ".vuepress/client.mjs")
	].find((item) => fs.pathExistsSync(item));
	return userConfigPlugin;
};
//#endregion
//#region src/config/resolveConfig.ts
const resolveConfig = async ({ sourceDir, cliOptions, defaultAppConfig }) => {
	const cliAppConfig = resolveCliAppConfig(sourceDir, cliOptions);
	const { userConfig, userConfigDependencies, userConfigPath } = await resolveUserConfig({
		cliOptionsConfig: cliOptions.config,
		cliAppConfigSource: cliAppConfig.source
	});
	const userConfigPlugin = transformUserConfigToPlugin(userConfig, cliAppConfig.source);
	return {
		appConfig: resolveAppConfig({
			defaultAppConfig,
			cliAppConfig,
			userConfig
		}),
		userConfigDependencies,
		userConfigPath,
		userConfigPlugin
	};
};
//#endregion
//#region src/commands/build.ts
const log$1 = debug("vuepress:cli/build");
const build = async ({ defaultAppConfig, sourceDir = ".", cliOptions = {} }) => {
	const start = Date.now();
	log$1(`cliOptions:`, cliOptions);
	process.env.NODE_ENV ??= "production";
	const { appConfig, userConfigPlugin } = await resolveConfig({
		defaultAppConfig,
		sourceDir,
		cliOptions
	});
	if (appConfig === null) throw logger.createError("Failed to resolve config");
	const app = createBuildApp(appConfig);
	app.use(userConfigPlugin);
	if (cliOptions.cleanTemp === true) await withSpinner("Cleaning temp")(async () => {
		await fs.remove(app.dir.temp());
	});
	if (cliOptions.cleanCache === true) await withSpinner("Cleaning cache")(async () => {
		await fs.remove(app.dir.cache());
	});
	await fs.emptyDir(app.dir.dest());
	await withSpinner("Initializing and preparing data")(async () => {
		await app.init();
		await app.prepare();
	});
	await app.build();
	await app.pluginApi.hooks.onGenerated.process(app);
	logger.success(`VuePress build completed in ${formatMs(Date.now() - start)}!`);
};
//#endregion
//#region src/commands/dev/handlePageAdd.ts
/**
* Event handler for page add event
*
* Returns the added page
*/
const handlePageAdd = async (app, filePath) => {
	if (app.pages.findIndex((page) => page.filePath === filePath) !== -1) return null;
	const page = await createPage(app, { filePath });
	app.pages.push(page);
	app.pagesMap[filePath] = page;
	await preparePageChunk(app, page);
	await prepareRoutes(app);
	await app.pluginApi.hooks.onPageUpdated.process(app, "create", page, null);
	return page;
};
//#endregion
//#region src/commands/dev/handlePageChange.ts
/**
* Event handler for page change event
*
* Returns the old page and the new page tuple
*/
const handlePageChange = async (app, filePath) => {
	const pageIndex = app.pages.findIndex((page) => page.filePath === filePath);
	if (pageIndex === -1) return null;
	const pageOld = app.pages[pageIndex];
	const pageNew = await createPage(app, { filePath });
	app.pages.splice(pageIndex, 1, pageNew);
	app.pagesMap[filePath] = pageNew;
	await preparePageChunk(app, pageNew);
	const isPathChanged = pageOld.path !== pageNew.path;
	const isRouteMetaChanged = JSON.stringify(pageOld.routeMeta) !== JSON.stringify(pageNew.routeMeta);
	if (isPathChanged || isRouteMetaChanged) await prepareRoutes(app);
	await app.pluginApi.hooks.onPageUpdated.process(app, "update", pageNew, pageOld);
	return [pageOld, pageNew];
};
//#endregion
//#region src/commands/dev/handlePageUnlink.ts
/**
* Event handler for page unlink event
*
* Returns the removed page
*/
const handlePageUnlink = async (app, filePath) => {
	const pageIndex = app.pages.findIndex((page) => page.filePath === filePath);
	if (pageIndex === -1) return null;
	const page = app.pages[pageIndex];
	app.pages.splice(pageIndex, 1);
	delete app.pagesMap[filePath];
	await prepareRoutes(app);
	await app.pluginApi.hooks.onPageUpdated.process(app, "delete", null, page);
	return page;
};
//#endregion
//#region src/commands/dev/pageDepsHelper.ts
/**
* Create page deps helper
*/
const createPageDepsHelper = () => {
	const store = /* @__PURE__ */ new Map();
	return {
		add: ({ deps, filePathRelative }) => {
			const depsAdded = [];
			if (filePathRelative) deps.forEach((item) => {
				if (!store.has(item)) {
					store.set(item, /* @__PURE__ */ new Set());
					depsAdded.push(item);
				}
				store.get(item)?.add(filePathRelative);
			});
			return depsAdded;
		},
		remove: ({ deps, filePathRelative }) => {
			const depsRemoved = [];
			if (filePathRelative) deps.forEach((item) => {
				const pagePathsSet = store.get(item);
				pagePathsSet?.delete(filePathRelative);
				if (pagePathsSet?.size === 0) {
					store.delete(item);
					depsRemoved.push(item);
				}
			});
			return depsRemoved;
		},
		get: (dep) => {
			const pagePathsSet = store.get(dep);
			return pagePathsSet ? [...pagePathsSet] : [];
		}
	};
};
//#endregion
//#region src/commands/dev/processPagePatterns.ts
/**
* Process page patterns into match and ignore patterns.
*
* - Separates negated patterns (prefixed with `!`) from positive patterns.
* - Merges `DEFAULT_IGNORE_PATTERNS` into ignore patterns.
* - Expands directory patterns with `/**` suffix for picomatch compatibility,
*   since picomatch does not auto-expand like tinyglobby's `expandDirectories`.
*/
const processPagePatterns = (pagePatterns) => {
	const matchPatterns = [];
	const ignorePatterns = [...DEFAULT_IGNORE_PATTERNS];
	for (const pattern of pagePatterns) if (pattern.startsWith("!") && pattern[1] !== "(") ignorePatterns.push(pattern.slice(1));
	else matchPatterns.push(pattern);
	return {
		matchPatterns,
		ignorePatterns: ignorePatterns.flatMap((p) => {
			if (p.endsWith("/**")) return [p.slice(0, -3), p];
			return p.endsWith("**") ? [p] : [p, `${p}/**`];
		})
	};
};
//#endregion
//#region src/commands/dev/watchPageFiles.ts
/**
* Watch page files and deps, return file watchers
*/
const watchPageFiles = (app) => {
	const depsWatcher = chokidar.watch([], { ignoreInitial: true });
	const depsHelper = createPageDepsHelper();
	const addDeps = (page) => {
		const depsToAdd = depsHelper.add(page);
		depsWatcher.add(depsToAdd);
	};
	const removeDeps = (page) => {
		const depsToRemove = depsHelper.remove(page);
		depsWatcher.unwatch(depsToRemove);
	};
	const depsListener = async (dep) => {
		const pagePaths = depsHelper.get(dep);
		for (const filePathRelative of pagePaths) {
			logger.info(`dependency of page ${colors.magenta(filePathRelative)} is modified`);
			await handlePageChange(app, app.dir.source(filePathRelative));
		}
	};
	depsWatcher.on("add", depsListener);
	depsWatcher.on("change", depsListener);
	depsWatcher.on("unlink", depsListener);
	app.pages.forEach((page) => {
		addDeps(page);
	});
	const { matchPatterns, ignorePatterns } = processPagePatterns(app.options.pagePatterns);
	const sourceDir = app.dir.source();
	const tempDir = app.dir.temp();
	const cacheDir = app.dir.cache();
	const ignoreMatcher = picomatch(ignorePatterns);
	const pageMatcher = picomatch(matchPatterns, { ignore: ignorePatterns });
	const pagesWatcher = chokidar.watch(".", {
		cwd: sourceDir,
		ignored: (filepath, stats) => {
			const relative = path.relative(sourceDir, filepath);
			if (ignoreMatcher(relative)) return true;
			if (filepath === tempDir || filepath === cacheDir) return true;
			return !!stats?.isFile() && !pageMatcher(relative);
		},
		ignoreInitial: true
	});
	pagesWatcher.on("add", async (filePathRelative) => {
		logger.info(`page ${colors.magenta(filePathRelative)} is created`);
		const page = await handlePageAdd(app, app.dir.source(filePathRelative));
		if (page === null) return;
		addDeps(page);
	});
	pagesWatcher.on("change", async (filePathRelative) => {
		logger.info(`page ${colors.magenta(filePathRelative)} is modified`);
		const result = await handlePageChange(app, app.dir.source(filePathRelative));
		if (result === null) return;
		const [pageOld, pageNew] = result;
		removeDeps(pageOld);
		addDeps(pageNew);
	});
	pagesWatcher.on("unlink", async (filePathRelative) => {
		logger.info(`page ${colors.magenta(filePathRelative)} is removed`);
		const page = await handlePageUnlink(app, app.dir.source(filePathRelative));
		if (page === null) return;
		removeDeps(page);
	});
	return [pagesWatcher, depsWatcher];
};
//#endregion
//#region src/commands/dev/watchUserConfigFile.ts
const watchUserConfigFile = ({ userConfigPath, userConfigDependencies, restart }) => {
	const cwd = process.cwd();
	const configWatcher = chokidar.watch(userConfigPath, {
		cwd,
		ignoreInitial: true
	});
	configWatcher.on("change", (configFile) => {
		logger.info(`config ${colors.magenta(configFile)} is modified`);
		restart();
	});
	const depsWatcher = chokidar.watch(userConfigDependencies, {
		cwd,
		ignoreInitial: true
	});
	depsWatcher.on("change", (depFile) => {
		logger.info(`config dependency ${colors.magenta(depFile)} is modified`);
		restart();
	});
	return [configWatcher, depsWatcher];
};
//#endregion
//#region src/commands/dev/dev.ts
const log = debug("vuepress:cli/dev");
const dev = async ({ defaultAppConfig, sourceDir = ".", cliOptions = {} }) => {
	log(`cliOptions:`, cliOptions);
	process.env.NODE_ENV ??= "development";
	const { appConfig, userConfigDependencies, userConfigPath, userConfigPlugin } = await resolveConfig({
		defaultAppConfig,
		sourceDir,
		cliOptions
	});
	if (appConfig === null) throw logger.createError("Failed to resolve config");
	const app = createDevApp(appConfig);
	app.use(userConfigPlugin);
	if (cliOptions.cleanTemp === true) await withSpinner("Cleaning temp")(async () => fs.remove(app.dir.temp()));
	if (cliOptions.cleanCache === true) await withSpinner("Cleaning cache")(async () => fs.remove(app.dir.cache()));
	await withSpinner("Initializing and preparing data")(async () => {
		await app.init();
		await app.prepare();
	});
	const close = await app.dev();
	if (cliOptions.watch === false) return;
	const watchers = [];
	const restart = async () => {
		await Promise.all([...watchers.map(async (item) => item.close()), close()]);
		app.writeTemp.cleanup();
		await dev({
			defaultAppConfig,
			sourceDir,
			cliOptions: {
				...cliOptions,
				cleanCache: false,
				cleanTemp: false
			}
		});
		logger.tip(`dev server has restarted, please refresh your browser`);
	};
	watchers.push(...watchPageFiles(app));
	if (userConfigPath) watchers.push(...watchUserConfigFile({
		userConfigPath,
		userConfigDependencies,
		restart
	}));
	await app.pluginApi.hooks.onWatched.process(app, watchers, restart);
};
//#endregion
//#region src/commands/info.ts
const info = async () => {
	const spinner = ora();
	spinner.start("Collecting Environment Info");
	const result = await envinfo.run({
		System: [
			"OS",
			"CPU",
			"Memory",
			"Shell"
		],
		Binaries: [
			"bun",
			"Node",
			"npm",
			"pnpm",
			"Yarn"
		],
		Utilities: ["Git"],
		Browsers: [
			"Chrome",
			"Edge",
			"Firefox",
			"Safari"
		],
		npmPackages: [
			"@vuepress/bundler-vite",
			"@vuepress/bundler-webpack",
			"@vuepress/cli",
			"@vuepress/client",
			"@vuepress/core",
			"@vuepress/markdown",
			"@vuepress/shared",
			"@vuepress/utils",
			"vuepress",
			"vue",
			"vue-router"
		]
	}, {
		showNotFound: true,
		duplicates: true,
		fullTree: true
	});
	spinner.stop();
	logger.info(result);
};
//#endregion
//#region src/cli.ts
const require = createRequire(import.meta.url);
/**
* Vuepress cli
*/
const cli = (defaultAppConfig = {}) => {
	const program = cac("vuepress");
	const versionCli = require("../package.json").version;
	const versionCore = require("@vuepress/core/package.json").version;
	program.version(`core@${versionCore} vuepress/cli@${versionCli}`);
	program.help();
	program.command("dev [sourceDir]", "Start development server").option("-c, --config <config>", "Set path to config file").option("-p, --port <port>", "Use specified port (default: 8080)").option("-t, --temp <temp>", "Set the directory of the temporary files").option("--host <host>", "Use specified host (default: 0.0.0.0)").option("--cache <cache>", "Set the directory of the cache files").option("--clean-temp", "Clean the temporary files before dev").option("--clean-cache", "Clean the cache files before dev").option("--open", "Open browser when ready").option("--debug", "Enable debug mode").option("--no-watch", "Disable watching page and config files").action(async (sourceDir, cliOptions) => dev({
		defaultAppConfig,
		sourceDir,
		cliOptions
	}));
	program.command("build [sourceDir]", "Build to static site").option("-c, --config <config>", "Set path to config file").option("-d, --dest <dest>", "Set the directory build output (default: .vuepress/dist)").option("-t, --temp <temp>", "Set the directory of the temporary files").option("--cache <cache>", "Set the directory of the cache files").option("--clean-temp", "Clean the temporary files before build").option("--clean-cache", "Clean the cache files before build").option("--debug", "Enable debug mode").action(async (sourceDir, cliOptions) => build({
		defaultAppConfig,
		sourceDir,
		cliOptions
	}));
	program.command("info", "Display environment information").action(info);
	program.parse(process.argv, { run: false });
	if (program.matchedCommand) program.runMatchedCommand().catch((err) => {
		logger.error(err instanceof Error ? colors.red(err.stack) : err);
		process.exit(1);
	});
	else program.outputHelp();
};
//#endregion
//#region src/defineUserConfig.ts
const defineUserConfig = (config) => config;
//#endregion
export { build, cli, createPageDepsHelper, defineUserConfig, dev, handlePageAdd, handlePageChange, handlePageUnlink, info, loadUserConfig, processPagePatterns, resolveAppConfig, resolveCliAppConfig, resolveConfig, resolveUserConfig, resolveUserConfigConventionalPath, resolveUserConfigPath, transformUserConfigToPlugin, watchPageFiles, watchUserConfigFile };
