// @vuepress/core/dist/index.js
import { createRequire } from "node:module";
import { ensureLeadingSlash, formatDateString, inferRoutePath, isFunction, isString, normalizeRoutePath, omit, removeLeadingSlash, resolveLocalePath } from "@vuepress/shared";
import { colors, debug, fs, hash, logger, path, sanitizeFileName, templateRenderer, tinyglobby, transformPathToFileName, warn } from "@vuepress/utils";
import { createMarkdown } from "@vuepress/markdown";
//#region src/app/prepare/prepareClientConfigs.ts
/**
* Generate client configs temp file
*
* @internal
*/
const prepareClientConfigs = async (app) => {
	const clientConfigFiles = await app.pluginApi.hooks.clientConfigFile.process(app);
	const content = `\
${clientConfigFiles.map((filePath, index) => `import * as clientConfig${index} from '${filePath}'`).join("\n")}

export const clientConfigs = [
${clientConfigFiles.map((_, index) => `  clientConfig${index},`).join("\n")}
].map((m) => m.default).filter(Boolean)
`;
	await app.writeTemp("internal/clientConfigs.js", content);
};
//#endregion
//#region src/page/inferPagePath.ts
/**
* Infer page path according to file path
*
* @internal
*/
const inferPagePath = ({ app, filePathRelative, options }) => {
	if (options.path) {
		const pathLocale = resolveLocalePath(app.siteData.locales, options.path);
		return {
			pathInferred: filePathRelative ? inferRoutePath(ensureLeadingSlash(filePathRelative)) : null,
			pathLocale
		};
	}
	if (!filePathRelative) return {
		pathInferred: null,
		pathLocale: "/"
	};
	const pathInferred = inferRoutePath(ensureLeadingSlash(filePathRelative));
	return {
		pathInferred,
		pathLocale: resolveLocalePath(app.siteData.locales, pathInferred)
	};
};
//#endregion
//#region src/page/parsePageContent.ts
/**
* Render page content and extract related info
*
* @internal
*/
const parsePageContent = ({ app, content, filePath, filePathRelative, options }) => {
	const markdownEnv = {
		base: app.options.base,
		filePath,
		filePathRelative,
		frontmatter: { ...options.frontmatter }
	};
	const contentRendered = app.markdown.render(content, markdownEnv);
	/* istanbul ignore next */
	const { frontmatter = {}, headers = [], importedFiles = [], links = [], sfcBlocks = {
		template: null,
		script: null,
		scriptSetup: null,
		scripts: [],
		styles: [],
		customBlocks: []
	}, title = "", ...extraMarkdownEnv } = markdownEnv;
	return {
		contentRendered,
		deps: importedFiles,
		frontmatter,
		headers,
		links,
		markdownEnv: omit(extraMarkdownEnv, "base", "content", "filePath", "filePathRelative", "frontmatter"),
		sfcBlocks,
		title: frontmatter.title ? `${frontmatter.title}` : title
	};
};
//#endregion
//#region src/page/resolvePageChunkInfo.ts
/**
* Resolve page chunk file related info
*
* @internal
*/
const resolvePageChunkInfo = ({ app, filePath, filePathRelative, htmlFilePathRelative }) => {
	const chunkName = transformPathToFileName(htmlFilePathRelative);
	if (filePath && filePathRelative) return {
		chunkFilePath: filePath,
		chunkFilePathRelative: filePathRelative,
		chunkName
	};
	const chunkFilePathRelative = path.join("pages", `${htmlFilePathRelative}.vue`);
	return {
		chunkFilePath: app.dir.temp(chunkFilePathRelative),
		chunkFilePathRelative,
		chunkName
	};
};
//#endregion
//#region src/page/resolvePageContent.ts
const log$5 = debug("vuepress:core/page");
const FALLBACK_CONTENT = "";
/**
* Resolve page content according to `content` or `filePath`
*
* @internal
*/
const resolvePageContent = async ({ filePath, options }) => {
	if (isString(options.content)) return options.content;
	if (filePath) try {
		return await fs.readFile(filePath, "utf-8");
	} catch (e) {
		log$5(e instanceof Error ? e.message : e);
	}
	return FALLBACK_CONTENT;
};
//#endregion
//#region src/page/resolvePageDate.ts
const FILENAME_DATE_RE = /^(\d{4})-(\d{1,2})(?:-(\d{1,2}))?-(.*)$/;
const DIRNAME_DATE_RE = /(\d{4})\/(\d{1,2})(?:\/(\d{1,2}))?(\/|$)/;
const DEFAULT_DATE = "0000-00-00";
/**
* Resolve page date according to frontmatter or file path
*
* It will be resolved as 'yyyy-MM-dd' format
*
* @internal
*/
const resolvePageDate = ({ frontmatter, filePathRelative }) => {
	if (frontmatter.date instanceof Date) return formatDateString([
		frontmatter.date.getUTCFullYear(),
		frontmatter.date.getUTCMonth() + 1,
		frontmatter.date.getUTCDate()
	].join("-"), DEFAULT_DATE);
	if (isString(frontmatter.date)) return formatDateString(frontmatter.date, DEFAULT_DATE);
	if (filePathRelative === null) return DEFAULT_DATE;
	const filename = path.parse(filePathRelative).name;
	if (filename) {
		const matches = filename.match(FILENAME_DATE_RE);
		if (matches) return formatDateString(`${matches[1]}-${matches[2]}-${matches[3] ?? "01"}`, DEFAULT_DATE);
	}
	const dirname = path.dirname(filePathRelative);
	if (dirname !== ".") {
		const matches = dirname.match(DIRNAME_DATE_RE);
		if (matches) return formatDateString(`${matches[1]}-${matches[2]}-${matches[3] ?? "01"}`, DEFAULT_DATE);
	}
	return DEFAULT_DATE;
};
//#endregion
//#region src/page/resolvePageFilePath.ts
/**
* Resolve absolute and relative path of page file
*
* @internal
*/
const resolvePageFilePath = ({ app, options }) => {
	if (!options.filePath) return {
		filePath: null,
		filePathRelative: null
	};
	if (!path.isAbsolute(options.filePath)) throw logger.createError(`filePath is not absolute file path: ${options.filePath}}`);
	return {
		filePath: options.filePath,
		filePathRelative: path.relative(app.dir.source(), options.filePath)
	};
};
//#endregion
//#region src/page/resolvePageHtmlInfo.ts
/**
* Resolve page rendered html file path
*
* @internal
*/
const resolvePageHtmlInfo = ({ app, path: pagePath }) => {
	const path = decodeURI(pagePath);
	const htmlFilePathRelative = removeLeadingSlash(path.endsWith("/") ? `${path}index.html` : path.endsWith(".html") ? path : `${path}.html`);
	return {
		htmlFilePath: app.dir.dest(htmlFilePathRelative),
		htmlFilePathRelative
	};
};
//#endregion
//#region src/page/resolvePageLang.ts
/**
* Resolve language of page
*
* @internal
*/
const resolvePageLang = ({ app, frontmatter, pathLocale }) => {
	if (isString(frontmatter.lang) && frontmatter.lang) return frontmatter.lang;
	return app.siteData.locales[pathLocale]?.lang ?? app.siteData.lang;
};
//#endregion
//#region src/page/resolvePagePath.ts
/**
* Resolve the final route path of a page
*
* @internal
*/
const resolvePagePath = ({ permalink, pathInferred, options }) => {
	const pagePath = options.path || permalink || pathInferred;
	if (!pagePath) throw logger.createError(`page path is empty, page options: ${JSON.stringify(options, null, 2)}`);
	return encodeURI(pagePath.split("/").map(sanitizeFileName).join("/"));
};
//#endregion
//#region src/page/resolvePagePermalink.ts
/**
* Resolve page permalink from frontmatter / options / pattern
*
* @internal
*/
const resolvePagePermalink = ({ app, frontmatter, slug, date, pathInferred, pathLocale }) => {
	if (frontmatter.permalink === null) return null;
	if (isString(frontmatter.permalink)) return frontmatter.permalink;
	if (frontmatter.permalinkPattern === null) return null;
	const permalinkPattern = frontmatter.permalinkPattern || app.options.permalinkPattern;
	if (!isString(permalinkPattern)) return null;
	const [year, month, day] = date.split("-");
	return ensureLeadingSlash(path.join(pathLocale, permalinkPattern.replace(/:year/, year).replace(/:month/, month).replace(/:day/, day).replace(/:slug/, slug).replace(/:raw/, pathInferred?.replace(/^\//, "") ?? "")));
};
//#endregion
//#region src/page/resolvePageRouteMeta.ts
/**
* Resolve page route meta
*
* @internal
*/
const resolvePageRouteMeta = ({ frontmatter }) => frontmatter.routeMeta ?? {};
//#endregion
//#region src/page/resolvePageSlug.ts
const DATE_RE = /(\d{4}-\d{1,2}(-\d{1,2})?)-(.*)/;
/**
* Resolve page slug from filename
*
* @internal
*/
const resolvePageSlug = ({ filePathRelative }) => {
	if (!filePathRelative) return "";
	const filename = path.parse(filePathRelative).name;
	const match = filename.match(DATE_RE);
	return match ? match[3] : filename;
};
//#endregion
//#region src/page/createPage.ts
/**
* Create vuepress page object
*/
const createPage = async (app, options) => {
	await app.pluginApi.hooks.extendsPageOptions.process(options, app);
	const { filePath, filePathRelative } = resolvePageFilePath({
		app,
		options
	});
	const content = await resolvePageContent({
		filePath,
		options
	});
	const { contentRendered, deps, frontmatter, headers, links, markdownEnv, sfcBlocks, title } = parsePageContent({
		app,
		content,
		filePath,
		filePathRelative,
		options
	});
	const routeMeta = resolvePageRouteMeta({ frontmatter });
	const slug = resolvePageSlug({ filePathRelative });
	const date = resolvePageDate({
		frontmatter,
		filePathRelative
	});
	const { pathInferred, pathLocale } = inferPagePath({
		app,
		filePathRelative,
		options
	});
	const lang = resolvePageLang({
		app,
		frontmatter,
		pathLocale
	});
	const permalink = resolvePagePermalink({
		app,
		frontmatter,
		slug,
		date,
		pathInferred,
		pathLocale
	});
	const path = resolvePagePath({
		permalink,
		pathInferred,
		options
	});
	const { htmlFilePath, htmlFilePathRelative } = resolvePageHtmlInfo({
		app,
		path
	});
	const { chunkFilePath, chunkFilePathRelative, chunkName } = resolvePageChunkInfo({
		app,
		filePath,
		filePathRelative,
		htmlFilePathRelative
	});
	const page = {
		data: {
			path,
			title,
			lang,
			frontmatter
		},
		path,
		title,
		lang,
		frontmatter,
		headers,
		content,
		contentRendered,
		date,
		deps,
		links,
		markdownEnv,
		pathInferred,
		pathLocale,
		permalink,
		routeMeta,
		sfcBlocks,
		slug,
		filePath,
		filePathRelative,
		chunkFilePath,
		chunkFilePathRelative,
		chunkName,
		htmlFilePath,
		htmlFilePathRelative
	};
	await app.pluginApi.hooks.extendsPage.process(page, app);
	return page;
};
//#endregion
//#region src/page/renderPageToVue.ts
const TEMPLATE_WRAPPER_TAG_OPEN = "<div>";
const TEMPLATE_WRAPPER_TAG_CLOSE = "</div>";
const SCRIPT_TAG_OPEN = "<script>";
const SCRIPT_TAG_CLOSE = "<\/script>";
const SCRIPT_TAG_OPEN_LANG_TS_REGEX = /<\s*script[^>]*\blang=['"]ts['"][^>]*/;
const SCRIPT_TAG_OPEN_LANG_TS = "<script lang=\"ts\">";
const SCRIPT_DEFAULT_EXPORT_REGEX = /((?:^|\n|;)\s*)export(\s*)default/;
const SCRIPT_DEFAULT_NAMED_EXPORT_REGEX = /((?:^|\n|;)\s*)export(.+)as(\s*)default/;
const SCRIPT_DEFAULT_EXPORT_CODE_TEMPLATE_OUTLET = "__SCRIPT_DEFAULT_EXPORT__";
const SCRIPT_DEFAULT_EXPORT_CODE_TEMPLATE = `export default { name: ${SCRIPT_DEFAULT_EXPORT_CODE_TEMPLATE_OUTLET} }`;
const PAGE_DATA_CODE_VAR_NAME = "_pageData";
const PAGE_DATA_CODE_TEMPLATE_OUTLET = "__PAGE_DATA__";
const PAGE_DATA_CODE_TEMPLATE = `export const ${PAGE_DATA_CODE_VAR_NAME} = JSON.parse(${PAGE_DATA_CODE_TEMPLATE_OUTLET})`;
const HMR_CODE$2 = `
if (import.meta.webpackHot) {
  import.meta.webpackHot.accept()
  __VUE_HMR_RUNTIME__.updatePageData?.(${PAGE_DATA_CODE_VAR_NAME})
}

if (import.meta.hot) {
  import.meta.hot.accept((m) => {
    __VUE_HMR_RUNTIME__.updatePageData?.(m.${PAGE_DATA_CODE_VAR_NAME})
  })
}
`;
/**
* Util to resolve the open tag of script block
*/
const resolveScriptTagOpen = (page) => {
	if (page.sfcBlocks.script?.tagOpen) return page.sfcBlocks.script.tagOpen;
	return page.sfcBlocks.scriptSetup?.tagOpen.match(SCRIPT_TAG_OPEN_LANG_TS_REGEX) ? SCRIPT_TAG_OPEN_LANG_TS : SCRIPT_TAG_OPEN;
};
/**
* Util to resolve the default export code
*/
const resolveDefaultExportCode = (page) => SCRIPT_DEFAULT_EXPORT_CODE_TEMPLATE.replace(SCRIPT_DEFAULT_EXPORT_CODE_TEMPLATE_OUTLET, JSON.stringify(path.basename(page.chunkFilePath)));
/**
* Util to resolve the page data code
*/
const resolvePageDataCode = (page) => PAGE_DATA_CODE_TEMPLATE.replace(PAGE_DATA_CODE_TEMPLATE_OUTLET, JSON.stringify(JSON.stringify(page.data)));
/**
* Resolve the stripped content of script block
*/
const resolveScriptContentStripped = (app, page) => {
	const rawContentStripped = page.sfcBlocks.script?.contentStripped;
	const hasDefaultExport = rawContentStripped ? SCRIPT_DEFAULT_EXPORT_REGEX.test(rawContentStripped) || SCRIPT_DEFAULT_NAMED_EXPORT_REGEX.test(rawContentStripped) : false;
	return [
		rawContentStripped,
		resolvePageDataCode(page),
		!hasDefaultExport && resolveDefaultExportCode(page),
		app.env.isDev && HMR_CODE$2
	].filter(isString).join("\n");
};
/**
* Render page to vue component
*/
const renderPageToVue = (app, page) => {
	const { sfcBlocks } = page;
	return [
		sfcBlocks.template && [
			sfcBlocks.template.tagOpen,
			TEMPLATE_WRAPPER_TAG_OPEN,
			sfcBlocks.template.contentStripped,
			TEMPLATE_WRAPPER_TAG_CLOSE,
			sfcBlocks.template.tagClose
		].join(""),
		[
			resolveScriptTagOpen(page),
			resolveScriptContentStripped(app, page),
			sfcBlocks.script?.tagClose ?? SCRIPT_TAG_CLOSE
		].filter(isString).join("\n"),
		sfcBlocks.scriptSetup?.content,
		...sfcBlocks.styles.map((item) => item.content),
		...sfcBlocks.customBlocks.map((item) => item.content)
	].filter(isString).join("\n");
};
//#endregion
//#region src/app/prepare/preparePageChunk.ts
/**
* Generate temp file if the page does not have a source file
*/
const preparePageChunk = async (app, page) => {
	if (page.filePath === null) await app.writeTemp(page.chunkFilePathRelative, renderPageToVue(app, page));
};
//#endregion
//#region src/app/prepare/prepareRoutes.ts
const ROUTES_VAR_NAME = "routes";
const REDIRECTS_VAR_NAME = "redirects";
const HMR_CODE$1 = `
if (import.meta.webpackHot) {
  import.meta.webpackHot.accept()
  __VUE_HMR_RUNTIME__.updateRoutes?.(${ROUTES_VAR_NAME})
  __VUE_HMR_RUNTIME__.updateRedirects?.(${REDIRECTS_VAR_NAME})
}

if (import.meta.hot) {
  import.meta.hot.accept((m) => {
    __VUE_HMR_RUNTIME__.updateRoutes?.(m.${ROUTES_VAR_NAME})
    __VUE_HMR_RUNTIME__.updateRedirects?.(m.${REDIRECTS_VAR_NAME})
  })
}
`;
/**
* Resolve page redirects
*
* @internal
*/
const resolvePageRedirects = ({ path, pathInferred }) => {
	const redirectsSet = /* @__PURE__ */ new Set();
	if (pathInferred !== null) {
		const normalizedPathInferred = normalizeRoutePath(pathInferred);
		const encodedPathInferred = encodeURI(normalizedPathInferred);
		if (normalizedPathInferred !== path && encodedPathInferred !== path) redirectsSet.add(encodedPathInferred);
	}
	return Array.from(redirectsSet);
};
/**
* Generate routes temp file
*/
const prepareRoutes = async (app) => {
	let content = `\
export const ${REDIRECTS_VAR_NAME} = JSON.parse(${JSON.stringify(JSON.stringify(Object.fromEntries(app.pages.flatMap((page) => resolvePageRedirects(page).map((redirect) => [redirect, page.path])))))})

export const ${ROUTES_VAR_NAME} = Object.fromEntries([
${app.pages.map(({ chunkFilePath, chunkName, path, routeMeta }) => `  [${JSON.stringify(path)}, { loader: () => import(${chunkName ? `/* webpackChunkName: "${chunkName}" */` : ""}${JSON.stringify(chunkFilePath)}), meta: ${JSON.stringify(routeMeta)} }],`).join("\n")}
]);
`;
	if (app.env.isDev) content += HMR_CODE$1;
	await app.writeTemp("internal/routes.js", content);
};
//#endregion
//#region src/app/prepare/prepareSiteData.ts
const SITE_DATA_VAR_NAME = "siteData";
const HMR_CODE = `
if (import.meta.webpackHot) {
  import.meta.webpackHot.accept()
  __VUE_HMR_RUNTIME__.updateSiteData?.(${SITE_DATA_VAR_NAME})
}

if (import.meta.hot) {
  import.meta.hot.accept((m) => {
    __VUE_HMR_RUNTIME__.updateSiteData?.(m.${SITE_DATA_VAR_NAME})
  })
}
`;
/**
* Generate site data temp file
*
* @internal
*/
const prepareSiteData = async (app) => {
	let content = `\
export const ${SITE_DATA_VAR_NAME} = JSON.parse(${JSON.stringify(JSON.stringify(app.siteData))})
`;
	if (app.env.isDev) content += HMR_CODE;
	await app.writeTemp("internal/siteData.js", content);
};
//#endregion
//#region src/app/resolveAppMarkdown.ts
/**
* Resolve markdown-it instance for vuepress app
*
* @internal
*/
const resolveAppMarkdown = async (app) => {
	await app.pluginApi.hooks.extendsMarkdownOptions.process(app.options.markdown, app);
	const markdown = createMarkdown(app.options.markdown);
	await app.pluginApi.hooks.extendsMarkdown.process(markdown, app);
	return markdown;
};
//#endregion
//#region src/constants.ts
const DEFAULT_IGNORE_PATTERNS = [
	"**/node_modules",
	"**/.yarn",
	"**/.git",
	"**/.svn",
	"**/.hg"
];
//#endregion
//#region src/app/resolveAppPages.ts
const log$4 = debug("vuepress:core/app");
/**
* Resolve pages for vuepress app
*
* @internal
*/
const resolveAppPages = async (app) => {
	log$4("resolveAppPages start");
	const pageFilePaths = await tinyglobby.glob(app.options.pagePatterns, {
		absolute: true,
		cwd: app.dir.source(),
		ignore: DEFAULT_IGNORE_PATTERNS
	});
	let hasNotFoundPage = false;
	const pagesMap = {};
	const pages = await Promise.all(pageFilePaths.map(async (filePath) => {
		const page = await createPage(app, { filePath });
		if (page.filePath) pagesMap[page.filePath] = page;
		if (page.path === "/404.html") {
			page.frontmatter.layout ??= "NotFound";
			hasNotFoundPage = true;
		}
		return page;
	}));
	if (!hasNotFoundPage) pages.push(await createPage(app, {
		path: "/404.html",
		frontmatter: { layout: "NotFound" },
		content: "404 Not Found"
	}));
	log$4("resolveAppPages finish");
	return {
		pages,
		pagesMap
	};
};
//#endregion
//#region src/app/appInit.ts
const log$3 = debug("vuepress:core/app");
/**
* Initialize a vuepress app
*
* Plugins should be used before initialization.
*
* @internal
*/
const appInit = async (app) => {
	log$3("init start");
	app.pluginApi.registerHooks();
	app.markdown = await resolveAppMarkdown(app);
	const { pages, pagesMap } = await resolveAppPages(app);
	app.pages = pages;
	app.pagesMap = pagesMap;
	await app.pluginApi.hooks.onInitialized.process(app);
	log$3("init finish");
};
//#endregion
//#region src/app/appPrepare.ts
const log$2 = debug("vuepress:core/app");
/**
* Prepare files for development or build
*
* - page chunks
* - routes
* - site data
* - other files that generated by plugins
*
* @internal
*/
const appPrepare = async (app) => {
	log$2("prepare start");
	await Promise.all([
		...app.pages.map(async (page) => preparePageChunk(app, page)),
		prepareRoutes(app),
		prepareSiteData(app),
		prepareClientConfigs(app)
	]);
	await app.pluginApi.hooks.onPrepared.process(app);
	log$2("prepare finish");
};
//#endregion
//#region src/app/resolvePluginObject.ts
/**
* Resolve a plugin object according to name / path / module and config
*
* @internal
*/
const resolvePluginObject = (app, plugin) => isFunction(plugin) ? plugin(app) : plugin;
//#endregion
//#region src/app/appUse.ts
const log$1 = debug("vuepress:core/app");
/**
* Use a plugin in vuepress app.
*
* Should be called before initialization.
*
* @internal
*/
const appUse = (app, rawPlugin) => {
	const pluginObject = resolvePluginObject(app, rawPlugin);
	if (!pluginObject.name) {
		warn(`an anonymous plugin or theme was detected and ignored`);
		return app;
	}
	log$1(`use plugin ${colors.magenta(pluginObject.name)}`);
	if (pluginObject.multiple !== true) {
		const duplicateIndex = app.pluginApi.plugins.findIndex(({ name }) => name === pluginObject.name);
		if (duplicateIndex !== -1) {
			app.pluginApi.plugins.splice(duplicateIndex, 1);
			warn(`plugin ${colors.magenta(pluginObject.name)} has been used multiple times, only the last one will take effect`);
		}
	}
	app.pluginApi.plugins.push(pluginObject);
	return app;
};
//#endregion
//#region src/pluginApi/createHookQueue.ts
const log = debug("vuepress:core/plugin-api");
/**
* Create hook queue for plugin system
*
* @internal
*/
const createHookQueue = (name) => {
	const items = [];
	return {
		name,
		items,
		add: (item) => {
			items.push(item);
		},
		process: async (...args) => {
			const results = [];
			for (const item of items) {
				log(`process ${colors.magenta(name)} from ${colors.magenta(item.pluginName)}`);
				try {
					const result = await item.hook(...args);
					if (result !== void 0) results.push(result);
				} catch (e) {
					logger.error(`error in hook ${colors.magenta(name)} from ${colors.magenta(item.pluginName)}`);
					throw e;
				}
			}
			return results;
		}
	};
};
//#endregion
//#region src/pluginApi/createPluginApiHooks.ts
/**
* Create hooks for plugin api
*
* @internal
*/
const createPluginApiHooks = () => ({
	onInitialized: createHookQueue("onInitialized"),
	onPrepared: createHookQueue("onPrepared"),
	onWatched: createHookQueue("onWatched"),
	onPageUpdated: createHookQueue("onPageUpdated"),
	onGenerated: createHookQueue("onGenerated"),
	extendsMarkdownOptions: createHookQueue("extendsMarkdownOptions"),
	extendsMarkdown: createHookQueue("extendsMarkdown"),
	extendsPageOptions: createHookQueue("extendsPageOptions"),
	extendsPage: createHookQueue("extendsPage"),
	extendsBundlerOptions: createHookQueue("extendsBundlerOptions"),
	clientConfigFile: createHookQueue("clientConfigFile"),
	alias: createHookQueue("alias"),
	define: createHookQueue("define")
});
//#endregion
//#region src/pluginApi/normalizeAliasDefineHook.ts
/**
* Normalize alias and define hook
*
* @internal
*/
const normalizeAliasDefineHook = (hook) => async (app, isServer) => isFunction(hook) ? hook(app, isServer) : hook;
//#endregion
//#region src/pluginApi/normalizeClientConfigFileHook.ts
/**
* Normalize hook for client config file
*
* @internal
*/
const normalizeClientConfigFileHook = (hook) => async (app) => {
	const clientConfigFileResult = isFunction(hook) ? await hook(app) : hook;
	if (!await fs.pathExists(clientConfigFileResult)) throw logger.createError(`client config file does not exist: ${clientConfigFileResult}`);
	return clientConfigFileResult;
};
//#endregion
//#region src/pluginApi/createPluginApiRegisterHooks.ts
/**
* Create registerHooks method for plugin api
*
* @internal
*/
const createPluginApiRegisterHooks = (plugins, hooks) => () => {
	plugins.forEach(({ name: pluginName, multiple, alias, define, clientConfigFile, ...commonHooks }) => {
		/**
		* hooks that need to be normalized
		*/
		if (alias) hooks.alias.add({
			pluginName,
			hook: normalizeAliasDefineHook(alias)
		});
		if (define) hooks.define.add({
			pluginName,
			hook: normalizeAliasDefineHook(define)
		});
		if (clientConfigFile) hooks.clientConfigFile.add({
			pluginName,
			hook: normalizeClientConfigFileHook(clientConfigFile)
		});
		/**
		* common hooks
		*/
		Object.keys(commonHooks).forEach((key) => {
			if (hooks[key] && commonHooks[key]) hooks[key].add({
				pluginName,
				hook: commonHooks[key]
			});
		});
	});
};
//#endregion
//#region src/pluginApi/createPluginApi.ts
/**
* Create vuepress plugin api
*
* @internal
*/
const createPluginApi = () => {
	const plugins = [];
	const hooks = createPluginApiHooks();
	return {
		plugins,
		hooks,
		registerHooks: createPluginApiRegisterHooks(plugins, hooks)
	};
};
//#endregion
//#region src/app/resolveAppDir.ts
const require$2 = createRequire(import.meta.url);
/**
* Create directory util function
*
* @internal
*/
const createAppDirFunction = (baseDir) => (...args) => path.resolve(baseDir, ...args);
/**
* Resolve directory utils for vuepress app
*/
const resolveAppDir = (options) => {
	const cache = createAppDirFunction(options.cache);
	const temp = createAppDirFunction(options.temp);
	const source = createAppDirFunction(options.source);
	const dest = createAppDirFunction(options.dest);
	const publicDir = createAppDirFunction(options.public);
	return {
		cache,
		temp,
		source,
		dest,
		client: createAppDirFunction(path.resolve(require$2.resolve("@vuepress/client/package.json"), "..")),
		public: publicDir
	};
};
//#endregion
//#region src/app/resolveAppEnv.ts
/**
* Resolve environment flags for vuepress app
*
* @internal
*/
const resolveAppEnv = (options) => ({
	isBuild: false,
	isDev: false,
	isDebug: options.debug
});
//#endregion
//#region src/app/resolveAppOptions.ts
const require$1 = createRequire(import.meta.url);
/**
* Create app options with default values
*
* @internal
*/
const resolveAppOptions = ({ base = "/", lang = "en-US", title = "", description = "", head = [], locales = {}, source, dest = path.resolve(source, ".vuepress/dist"), temp = path.resolve(source, ".vuepress/.temp"), cache = path.resolve(source, ".vuepress/.cache"), public: publicDir = path.resolve(source, ".vuepress/public"), host = "0.0.0.0", port = 8080, open = false, templateDev = path.normalize(require$1.resolve("@vuepress/client/templates/dev.html")), shouldPreload = true, shouldPrefetch = true, templateBuild = path.normalize(require$1.resolve("@vuepress/client/templates/build.html")), templateBuildRenderer = templateRenderer, bundler, debug = false, markdown = {}, pagePatterns = ["**/*.md", "!.vuepress"], permalinkPattern = null, plugins = [], theme }) => {
	locales["/"] ??= {
		lang,
		title,
		description
	};
	return {
		base,
		lang,
		title,
		description,
		head,
		locales,
		source,
		dest,
		temp,
		cache,
		public: publicDir,
		host,
		port,
		open,
		templateDev,
		shouldPreload,
		shouldPrefetch,
		templateBuild,
		templateBuildRenderer,
		bundler,
		debug,
		markdown,
		pagePatterns,
		permalinkPattern,
		plugins,
		theme
	};
};
//#endregion
//#region src/app/resolveAppSiteData.ts
/**
* Resolve site data for vuepress app
*
* Site data will also be used in client
*
* @internal
*/
const resolveAppSiteData = (options) => ({
	base: options.base,
	lang: options.lang,
	title: options.title,
	description: options.description,
	head: options.head,
	locales: options.locales
});
//#endregion
//#region src/app/resolveAppVersion.ts
const require = createRequire(import.meta.url);
/**
* Resolve version of vuepress app
*
* @internal
*/
const resolveAppVersion = () => {
	return fs.readJsonSync(require.resolve("@vuepress/core/package.json")).version;
};
//#endregion
//#region src/app/resolveAppWriteTemp.ts
/**
* Resolve write temp file util for vuepress app
*
* @internal
*/
const resolveAppWriteTemp = (dir) => {
	const cache = /* @__PURE__ */ new Map();
	const writeTemp = (async (file, content) => {
		const filePath = dir.temp(file);
		const contentHash = hash(content);
		let item = cache.get(filePath);
		if (!item) cache.set(filePath, item = {});
		if (item.hash === contentHash) return filePath;
		item.hash = contentHash;
		if (!item.current) item.current = (async () => {
			await fs.outputFile(filePath, content);
			item.current = item.next?.();
			return item.current;
		})();
		else item.next = async () => {
			await fs.outputFile(filePath, content);
			item.next = void 0;
			item.current = void 0;
		};
		await item.current;
		return filePath;
	});
	writeTemp.cleanup = () => {
		cache.clear();
	};
	return writeTemp;
};
//#endregion
//#region src/app/createBaseApp.ts
/**
* Create base vuepress app.
*
* Notice that the base app could not be used for dev nor build.
*
* It would be used for creating dev app or build app internally.
*
* @internal
*/
const createBaseApp = (config) => {
	const options = resolveAppOptions(config);
	const dir = resolveAppDir(options);
	const env = resolveAppEnv(options);
	const pluginApi = createPluginApi();
	const app = {
		options,
		siteData: resolveAppSiteData(options),
		version: resolveAppVersion(),
		dir,
		env,
		pluginApi,
		writeTemp: resolveAppWriteTemp(dir),
		use: (plugin) => appUse(app, plugin),
		init: async () => appInit(app),
		prepare: async () => appPrepare(app)
	};
	return app;
};
//#endregion
//#region src/app/resolveThemeInfo.ts
/**
* Resolve theme info and its parent theme info
*
* @internal
*/
const resolveThemeInfo = (app, theme) => {
	const themeObject = resolvePluginObject(app, theme);
	const themeInfo = {
		plugins: [...themeObject.plugins ?? [], themeObject],
		templateBuild: themeObject.templateBuild,
		templateBuildRenderer: themeObject.templateBuildRenderer,
		templateDev: themeObject.templateDev
	};
	if (!themeObject.extends) return themeInfo;
	const parentThemeInfo = resolveThemeInfo(app, themeObject.extends);
	return {
		plugins: [...parentThemeInfo.plugins, ...themeInfo.plugins],
		templateBuild: themeObject.templateBuild ?? parentThemeInfo.templateBuild,
		templateBuildRenderer: themeObject.templateBuildRenderer ?? parentThemeInfo.templateBuildRenderer,
		templateDev: themeObject.templateDev ?? parentThemeInfo.templateDev
	};
};
//#endregion
//#region src/app/setupAppThemeAndPlugins.ts
/**
* Setup theme and plugins for vuepress app
*
* @internal
*/
const setupAppThemeAndPlugins = (app, config) => {
	const themeInfo = resolveThemeInfo(app, app.options.theme);
	app.options.templateDev = config.templateDev ?? themeInfo.templateDev ?? app.options.templateDev;
	app.options.templateBuild = config.templateBuild ?? themeInfo.templateBuild ?? app.options.templateBuild;
	app.options.templateBuildRenderer = config.templateBuildRenderer ?? themeInfo.templateBuildRenderer ?? app.options.templateBuildRenderer;
	[...themeInfo.plugins, ...app.options.plugins].flat().forEach((plugin) => app.use(plugin));
};
//#endregion
//#region src/app/createBuildApp.ts
/**
* Create vuepress build app.
*/
const createBuildApp = (config) => {
	const app = createBaseApp(config);
	app.env.isBuild = true;
	app.build = async () => app.options.bundler.build(app);
	setupAppThemeAndPlugins(app, config);
	return app;
};
//#endregion
//#region src/app/createDevApp.ts
/**
* Create vuepress dev app.
*/
const createDevApp = (config) => {
	const app = createBaseApp(config);
	app.env.isDev = true;
	app.dev = async () => app.options.bundler.dev(app);
	setupAppThemeAndPlugins(app, config);
	return app;
};
//#endregion
export { DEFAULT_IGNORE_PATTERNS, appInit, appPrepare, appUse, createAppDirFunction, createBaseApp, createBuildApp, createDevApp, createHookQueue, createPage, createPluginApi, createPluginApiHooks, createPluginApiRegisterHooks, inferPagePath, normalizeAliasDefineHook, normalizeClientConfigFileHook, parsePageContent, prepareClientConfigs, preparePageChunk, prepareRoutes, prepareSiteData, renderPageToVue, resolveAppDir, resolveAppEnv, resolveAppOptions, resolveAppPages, resolveAppSiteData, resolveAppVersion, resolveAppWriteTemp, resolvePageChunkInfo, resolvePageContent, resolvePageDate, resolvePageFilePath, resolvePageHtmlInfo, resolvePageLang, resolvePagePath, resolvePagePermalink, resolvePageRouteMeta, resolvePageSlug, resolvePluginObject, resolveThemeInfo };
